boards: pjrc: Added Support for the Teensy CLI Tools soft reboot functionality

jbehrensnx opened the pull request:
This PR adds the support for the teensy cli tools soft reboot function by adding a bootloader functionality similar to the bossa one to the teensy boards and adding the ability to pass cli tool flags to the teensy runner.

This allows for flashing the board without triggering the programming pin or pressing the reset button.

jbehrensnx committed:
boards: pjrc: Added Teensy Bootloader Support

Added support for the teensy bootloader, allowing teensy_loader_cli to
flash the the teensy40 and teensy41 boards without triggering the reset
pin.

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

jbehrensnx committed:
west: runners: teensy: Add flags support

Added ability to set teensy_loader_cli flags from teenys runner.

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

nordicjm commented on boards/pjrc/teensy4/Kconfig.defconfig:
```diff
@@ -13,6 +13,33 @@ config BUILD_OUTPUT_HEX
 config DISK_DRIVER_SDMMC
 	default y if DISK_DRIVERS
 
+config BOOTLOADER_TEENSY
```
not a defconfig option, move to `Kconfig`

nordicjm commented on boards/pjrc/teensy4/Kconfig.defconfig:
```diff
@@ -13,6 +13,33 @@ config BUILD_OUTPUT_HEX
 
+config BOOTLOADER_TEENSY
+	bool "Teensy bootloader support"
+	select SERIAL
+	select USB_DEVICE_STACK
+	select USB_CDC_ACM
+	select CDC_ACM_DTE_RATE_CALLBACK_SUPPORT
+	select USB_DEVICE_INITIALIZE_AT_BOOT
```
@jfischer-no for review

nordicjm commented on boards/pjrc/teensy4/Kconfig.defconfig:
```diff
@@ -13,6 +13,33 @@ config BUILD_OUTPUT_HEX
+	help
+	  Sets wether the teensy bootloader should be loaded when an baudrate change is send or not.
+
+config BOOTLOADER_TEENSY_DEVICE_NAME
+	string "Teensy CDC ACM device name"
+	depends on BOOTLOADER_TEENSY
+	default "USB CDC-ACM"
+	help
+	  Sets the CDC ACM port to watch for reboot commands.
```
would also move

nordicjm commented on boards/pjrc/teensy4/bootloader_teensy.c:
```diff
@@ -0,0 +1,36 @@
+
+static void teensy_reset(const struct device *dev, uint32_t rate)
+{
+	if (rate != 134) {
```
what is 134?

nordicjm replied:
use defines/consts for these with descriptive names

nordicjm commented on boards/pjrc/teensy4/bootloader_teensy.c:
```diff
@@ -0,0 +1,36 @@
+	/* The programmer set the baud rate to 134 baud.  Reset into the
+	 * bootloader.
+	 */
+	__ASM volatile("bkpt #251");
```
and 251?

jbehrensnx replied:
`bkpt #251` is the assembly command to enter the bootloader.

nordicjm commented on boards/pjrc/teensy4/bootloader_teensy.c:
```diff
@@ -0,0 +1,36 @@
+	return cdc_acm_dte_rate_callback_set(dev, teensy_reset);
+}
+
+SYS_INIT(teensy_init, APPLICATION, 0);
```
use board hook

jbehrensnx commented:
The `cdc-acm-uart` device needed for this reset functionality is not by default included in the devicetree. My question is, would it make sense to add it in a disabled state? Users would have the choice to enable it or use a different usb device.

pdgendt commented:
@jbehrensnx PTAL the comments and CI issues

jbehrensnx commented:
I already implemented some of the feedback (through not pushed yet), but I am not sure when I will have time to work on this in the next couple of weeks.

jbehrensnx committed:
snippets: Added teensy bootloader snippet

Added a snippet that includes all Kconfig options and a dt overlay to
allow for easy use of the teensy bootloader.

Signed-off-by: Jan Behrens <jan.behrens@navimatix.de>

jbehrensnx commented:
Moved the bootloader to the soc section to better accommodate other boards, hopefully implemented all feedback and created a snippet for the bootloader, offloading some of the Kconfig configuration.

jfischer-no commented on snippets/pjrc-teensy-bootloader/pjrc-teensy-bootloader.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_USB_DEVICE_VID=0x16c0
+CONFIG_USB_DEVICE_PID=0x0483
+
+CONFIG_USB_DEVICE_STACK=y
+CONFIG_USB_CDC_ACM=y
+CONFIG_CDC_ACM_DTE_RATE_CALLBACK_SUPPORT=y
+CONFIG_USB_DEVICE_INITIALIZE_AT_BOOT=y
```
This is about to become deprecated.

jbehrensnx replied:
What would be the new way to do this?

jfischer-no commented on snippets/pjrc-teensy-bootloader/pjrc-teensy-bootloader.conf:
```diff
@@ -0,0 +1,9 @@
+CONFIG_BOOTLOADER_TEENSY=y
+CONFIG_USB_DEVICE_VID=0x16c0
+CONFIG_USB_DEVICE_PID=0x0483
```
We cannot use someone else's vendor ID in the tree.

jbehrensnx replied:
The thing is, it needs to be there, as the teensy cli looks for this specific VID/PID combination to send the soft-reboot command to.
I honestly have no idea what the process would be for getting permission to use the vendor id for this use case.

There are a handfull of uses of specific vendor ids in the tree, but I dont know their background.

Some other potential solutions, if we can not get permission to use the vendor id, would be to note the correct vendor id in the snippet readme, or look if we can get support for the zephyr vendor id in the teensy cli.

jbehrensnx replied:
Thanks for the clarification.

I did some investigation into teensy_cli, but the required changes are more difficult than they first appeared, so I currently don't plan on pursuing that avenue any further.

At this point it looks to me like the best option is to remove the vendor id from the snippet conf file and mention the correct vendor id in the readme instead.

I am also thinking about adding information about the snippet/soft reboot to the flashing section of the board readme.
